### PR TITLE
Send events to appservice based on room membership

### LINKED ISSUE #1679: Events should be forwarded to appservices who have members joined to the room
https://matrix.org/docs/spec/application_service/r0.1.2#registration

>Similarly, the application service is said to be interested in a given event if one of the application service's namespaced users is the target of the event, **or is a joined member of the room where the event occurred**.

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #1679 

~~Note that this does not currently work and needs input from maintainers~~ Thanks Neil!

### CODE REVIEW

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [inline appservice/consumers/roomserver.go:165]: This seems to be the wrong function to call. It expects a userId (presumably because this call is expected to handle `/members`). I'd like to just get membership of the room without the context of a sender user.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [inline appservice/consumers/roomserver.go:171]: Need help on this one also.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [inline appservice/consumers/roomserver.go:173]: I leave this up to the maintainers on the urgency of this, but obviously each event going through dendrite will currently check the membership of a room. If there is an easy way to cache *this* function call and invalidate it when the membership (or aliases) changes, then let's do that.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [inline appservice/consumers/roomserver.go:162]: I leave this up to the maintainers on the urgency of this, but obviously each event going through dendrite will currently check the membership of a room. If there is an easy way to cache this function call and invalidate it when the membership (or aliases) changes, then let's do that.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [reply appservice/consumers/roomserver.go]: We're using `QueryStateAfterEvents` which seems to do the right thing.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [reply appservice/consumers/roomserver.go:164]: Opened #1682 to track, we're consciously deferring this work as it's difficult to have caches span several components at once.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [inline appservice/consumers/roomserver.go:130]: > appservice/consumers/roomserver.go:130:1: cyclomatic complexity 15 of func `(*OutputRoomEventConsumer).appserviceIsInterestedInEvent` is high (> 13) (gocyclo)

Can I do something about this?

Comment (Half-Shot) [reply appservice/consumers/roomserver.go:159]: solved by splitting out to another function

File: appservice/consumers/roomserver.go
Comment (Kegsay) [inline appservice/consumers/roomserver.go:128]: `appserviceHasMembershipForEvent` or `appserviceHasMembershipInRoom`? Comments disagree.

**Review by neilalexander — APPROVED** [review]
Based on our chat yesterday I think this should be fine for the short term.

File: appservice/consumers/roomserver.go
Comment (Kegsay) [reply appservice/consumers/roomserver.go:128]: Also what does "has membership" mean? Joined? Invited? Left? They are all memberships...

File: appservice/consumers/roomserver.go
Comment (Kegsay) [inline appservice/consumers/roomserver.go:138]: This is the wrong API to use because we're pulling out the entire room state for this call. We can't make use of `QueryStateAfterEventsRequest.StateToFetch` because we want to glob all room member events. Use `QueryMembershipsForRoomRequest` instead:
```go
// QueryMembershipsForRoomRequest is a request to QueryMembershipsForRoom
type QueryMembershipsForRoomRequest struct {
	// If true, only returns the membership events of "join" membership
	JoinedOnly bool `json:"joined_only"`
	// ID of the room to fetch memberships from
	RoomID string `json:"room_id"`
	// ID of the user sending the request
	Sender string `json:"sender"`
}

// QueryMembershipsForRoomResponse is a response to QueryMembershipsForRoom
type QueryMembershipsForRoomResponse struct {
	// The "m.room.member" events (of "join" membership) in the client format
	JoinEvents []gomatrixserverlib.ClientEvent `json:"join_events"`
	// True if the user has been in room before and has either stayed in it or
	// left it.
	HasBeenInRoom bool `json:"has_been_in_room"`
	// True if the user asked to forget this room.
	IsRoomForgotten bool `json:"is_room_forgotten"`
}
```

**Review by Kegsay — CHANGES_REQUESTED** [review]
This is too heavy on memory at present.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [reply appservice/consumers/roomserver.go:138]: `QueryMembershipsForRoomRequest` doesn't work here because it takes a `sender` param, and we don't have a sender here.

File: appservice/consumers/roomserver.go
Comment (Half-Shot) [reply appservice/consumers/roomserver.go]: Changed function to reflect reality.

File: roomserver/internal/query/query.go
Comment (Kegsay) [inline roomserver/internal/query/query.go:245]: Please comment what the behaviour is supposed to be with an absent sender.

**Review by Kegsay — COMMENTED** [review]
It's still not great that we are pulling out a ton of `ClientEvent` structs for _every event_ we send to an AS, but this is marginally better than before.

File: roomserver/internal/query/query.go
Comment (neilalexander) [reply roomserver/internal/query/query.go:248]: Done, although I had already added a comment into the `api` package.

**Pull request merged**
